Time step and iteration indexation in pp.AD

vlipovac opened the pull request:
## Proposed changes

DRAFT: @IvarStefansson @keileg  for the indexation issues we discussed after the last user meeting.

Time step indexation reverted to start with 0 for first previous time step.
Time-dependent AD operators have `None` as their time step index, as long as they are at current time.
Iterative operators have always an integer value as iteration index, starting with 0 for the first previous iterate.
The first call to `.previous_iteration()` does not push the iterate index up (stays 0), but flags the instance as as previous iterate, removing it's capability to get a proper AdArray during parsing (`.values_and_jacobian()` will give a trivial `.jac`)


The "flag" for is previous iterate (also for time step) is computed using a PRIVATE index, which starts at -1, and is consistently increased with `+= 1`.

Note also, I propose here to return `iterate_index` as `None` for operators which are both `TimeDependentOperators` and `IterativeOperators`, and at a previous time step.

1. The aim is to make it more clear to users, that there are no iteration indices for previous time steps.
2. Also, code in various parts of the framework will be slightly shorter, because the `get_values` and `set_values` methods do not need an if-branch regarding whether the variable/operator is at previous time. Variables will always have a pair of time step index and iterate index, which are consistent with the intended usage of various get and set values methods in the Ad framework.


## Relevant changes

The essential changes are in the following files:

src/numerics/ad/_ad_utils.py
src/numerics/ad/equation_system.py
src/numerics/ad/operators.py

changes in the other file revert essentially the changes introduced in PR #1166

IvarStefansson commented on src/porepy/models/solution_strategy.py:
```diff
@@ -264,7 +264,7 @@ def time_step_indices(self) -> np.ndarray:
             An array of the indices of which time step solutions will be stored.
```
```suggestion
            An array of the indices of which time step solutions will be stored, counting
            from 0. Defaults to storing the most recently computed solution only.
```

IvarStefansson commented on src/porepy/numerics/ad/_ad_utils.py:
```diff
@@ -363,16 +350,20 @@ def set_solution_values(
+
+            Determines the key of where ``values`` are to be stored in
+            ``data[pp.TIME_STEP_SOLUTIONS][name]``.
+            0 is the first previous time step, 1 the one before that and so on.
```
```suggestion
            0 is the most recent time step, 1 the one before that and so on.
```

IvarStefansson commented on src/porepy/numerics/ad/_ad_utils.py:
```diff
@@ -409,31 +400,40 @@ def get_solution_values(
     time-dependent or iterative term.
 
+    Note:
+        Compared to :func:`get_solution_values` the getter works only for 1 defined
```
```suggestion
        Compared to :func:`set_solution_values` the getter works only for 1 defined
```

IvarStefansson commented on src/porepy/numerics/ad/_ad_utils.py:
```diff
@@ -409,31 +400,40 @@ def get_solution_values(
+
+            Determines the key of where ``values`` are to be stored in
+            ``data[pp.TIME_STEP_SOLUTIONS][name]``.
+            0 is the first previous time step, 1 the one before that and so on.
```
```suggestion
            0 is the most recent time step, 1 the one before that and so on.
```

IvarStefansson commented on src/porepy/numerics/ad/equation_system.py:
```diff
@@ -638,15 +639,15 @@ def set_variable_values(
-                iterative procedures.
+            variables: ``default=None``
+
+                VariableType input for which the values are requested.
```
```suggestion
                VariableType input for which the values are prescribed.
```

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1175,30 +1175,29 @@ def __init__(
+        else:
+            return self._time_step_index
 
     def previous_timestep(
```
What do you think about renaming these methods to the more precise `value_at_timestep/iteration(self, step)`?

IvarStefansson replied:
I believe you had some opinions on this, @keileg and @jwboth. Could you please comment?

IvarStefansson replied:
Works for me. What about the properties `is_previous_time` and `is_previous_iterate`? Do we get rid of "previous" there as well? Any suggestions for alternatives?

keileg replied:
> Works for me. What about the properties `is_previous_time` and `is_previous_iterate`? Do we get rid of "previous" there as well? Any suggestions for alternatives?

This is more difficult. I like the `is` prefix, indicating the boolean return type, but modifying the rest is difficult. One possibility is `is_lagged_in_time` (resp. `_iterate`), but that is not clearly better than using `previous` (so, switching to a synonym does not work). 

I think I'm blank, so unless other suggestions come up, I vote for keeping `is_previous_` (for now).

jwboth replied:
I have not much too add. I appreciate getting rid of previous when fetching values.

vlipovac replied:
`.previous_timestep(steps: int = 1) -> .at_timestep(steps: int = 1)`

Using *previous* reflects that the implementation is going *backwards* in time, with corresponding number of steps.
The naming `at_timestep` indicates that the user could use a concrete index as an argument to "jump" between time steps.
That functionality would require some more safety checks.

Analogous for iterates.

vlipovac replied:
Regarding the safety checks about stepping:

As of now, the user could create "holes" in the sequence of prev time and iter.
I.e. Operators for index 0, 1, 3,4 for example, but not for index 2.
Is this critical? Or does it fall in the category "The user should know what he is doing"?

IvarStefansson replied:
I don't understand what you mean about the "holes". Please elaborate why this could be an issue, in person if you prefer.

vlipovac replied:
Discussed in person, issue deemed not critical.

IvarStefansson replied:
Jakub and I suggest `is_at_timestep` for the booleans.

vlipovac replied:
I am not sure how that is interpreted as a boolean property. Are you inquiring for a specific time step?

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1285,26 +1284,41 @@ def __init__(
 
-        self._iterate_index: int = 0
+        self._iterate_index: int = -1
         """iterate index, starting with 0 (current iterate at current time) and
```
```suggestion
        """Iterate index, starting with 0 (current iterate at current time) and
```

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1285,26 +1284,41 @@ def __init__(
-        - 1 represents the first previous iterate
-        - 2 represents the iterate before that
+        - None indicates this instance is at a previous time
+        - 0 represents the first previous iterate
```
```suggestion
        - 0 represents the most recently computed iterate.
```

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1782,18 +1796,14 @@ def parse(self, mdg: pp.MixedDimensionalGrid) -> Any:
-                iterate_index=self.iterate_index,
-            )
+        # We can safely use both indices as arguments, without checking prev time,
+        # because iterate index is None if prev time, and vice versa
```
Thanks for the comment!

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1853,7 +1865,7 @@ def __init__(self, variables: list[Variable]) -> None:
             ), "Cannot create md-variable from variables at different time steps."
             assert (
-                len(set(iterate_indices)) == 1
+                len(set(iter_indices)) == 1 and len(set(current_iter)) == 1
```
Why the check of current_iter?

IvarStefansson replied:
Thanks for the explanation. Could you please add one line of comment.

vlipovac replied:
Documentation added.

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1864,16 +1876,24 @@ def __init__(self, variables: list[Variable]) -> None:
+
+        # If current time and iterate
+        if current_iter[0]:
+            self._iterate_index = -1  # current time and iter
```
In this case, isn't iter_indices[0] None?

vlipovac replied:
If all sub variables are current iterate, then `iter_indices[0] == 0`, because they fetch their `.val` from that index during `.value_and_jacobian()`.

vlipovac replied:
Discussed in person, added documentation why this is required and the if-else branch cannot be removed.

IvarStefansson commented on src/porepy/__init__.py:
```diff
@@ -160,6 +160,7 @@
 from porepy.numerics.ad.equation_system import EquationSystem
 from porepy.numerics.ad._ad_utils import set_solution_values
```
Could we merge these three lines to one?

vlipovac replied:
Can be merged into 1 import statement, but still spanning multiple lines because of the character limit.

vlipovac replied:
Merged into 1 import statement (for module `_ad_utils`) after discussion

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -440,9 +440,11 @@ def after_nonlinear_convergence(self) -> None:
         solution = self.equation_system.get_variable_values(iterate_index=0)
-        self.equation_system.shift_time_step_values()
+        self.equation_system.shift_time_step_values(
+            max_index=len(self.time_step_indices)
```
Added this functionality to avoid flooding the memory.
Same in line 430

IvarStefansson replied:
Seems like a very good idea!

IvarStefansson reviewed (commented):
Nice work. Please have a look at my comments and questions!

IvarStefansson commented on src/porepy/numerics/ad/operators.py:
```diff
@@ -1852,8 +1879,11 @@ def __init__(self, variables: list[Variable]) -> None:
             ), "Cannot create md-variable from variables at different time steps."
+            # NOTE both must be unique for all sub-variables, to avoid md-variables
+            # having sub-variables at different iterate states.
+            # Both current value, and most recent previous iterate have iterate index 0.
```
```suggestion
            # Both current value, and most recent previous iterate have iterate index 0,
            # hence the need to check the size of the current_iter set.
```

vlipovac commented on src/porepy/models/boundary_condition.py:
```diff
@@ -92,24 +95,22 @@ def update_boundary_condition(
+
+            # Before setting the new, most recent time step, shift the stored values
+            # backwards in time
+            pp.shift_solution_values(
```
I enhanced the functionality of the BoundaryConditions mixin, to now support the full time step depth.
Removed the warning below.

vlipovac commented:
@IvarStefansson @keileg @jwboth
I had a discussion with Ivar and I would like to finish this PR as it is stalling progress in the chain of PRs regarding CF.

Essentially, I took the liberty to rename the stuff, having your comments in mind.
The common denominator I was able to filter out was the wish to rename it to something resembling proper language and to add the word *at*.
I strongly discourage the use of *value* anywhere, since per logic we have two representations of the mathematical model: One is more symbolic (Operators and Operator trees), the other numerical (parsing operators and getting arrays).
I think we should be aware of these parallel structures, which of course go hand in hand, but their internal workings and functionality are orthogonal in some sense. This should be reflected in the naming convention. I.e., keywords like *value* should indicate a switch from the symbolic representation to the numeric evaluation.
The keyword *previous* I kept to indicate the directional behavior of the functionality, backwards in time and the iterate sense.

So the members are renamed as follows:

Functions creating operators at previous X

1. `previous_timestep()` -> `at_previous_timestep()`
2. `previous_iteration()` -> `at_previous_iteration()`

Boolean flags checking the index

1. `is_previous_timestep` -> `is_at_previous_timestep`
2. `is_previous_iterate` -> `is_at_previous_iteration`
3. `is_current_iterate` -> `is_at_current_iteration`

@IvarStefansson If you agree with this, please resolve the respective conversation and let's finalize this, but vennligst have a final thought on the changes here so we don't rush into pushing something confusing like in the previous PR.

jwboth commented:
> @IvarStefansson @keileg @jwboth I had a discussion with Ivar and I would like to finish this PR as it is stalling progress in the chain of PRs regarding CF.
> 
> Essentially, I took the liberty to rename the stuff, having your comments in mind. The common denominator I was able to filter out was the wish to rename it to something resembling proper language and to add the word _at_. I strongly discourage the use of _value_ anywhere, since per logic we have two representations of the mathematical model: One is more symbolic (Operators and Operator trees), the other numerical (parsing operators and getting arrays). I think we should be aware of these parallel structures, which of course go hand in hand, but their internal workings and functionality are orthogonal in some sense. This should be reflected in the naming convention. I.e., keywords like _value_ should indicate a switch from the symbolic representation to the numeric evaluation. The keyword _previous_ I kept to indicate the directional behavior of the functionality, backwards in time and the iterate sense.
> 
> So the members are renamed as follows:
> 
> Functions creating operators at previous X
> 
> 1. `previous_timestep()` -> `at_previous_time_step()`
> 2. `previous_iteration()` -> `at_previous_iteration()`
> 
> Boolean flags checking the index
> 
> 1. `is_previous_timestep` -> `is_at_previous_time_step`
> 2. `is_previous_iterate` -> `is_at_previous_iteration`
> 3. `is_current_iterate` -> `is_at_current_iteration`
> 
> The switch from `timestep` to `time_step` is made to be consistent with the `time_step_index`, which a property but also an argument in the equation system and ad utility functions.
> 
> @IvarStefansson If you agree with this, please resolve the respective conversation and let's finalize this, but vennligst have a final thought on the changes here so we don't rush into pushing something confusing like in the previous PR.

I give all authority to @IvarStefansson - I will accept whatever will be decided here. 

But since the final decision is not made yet, and before it is, here is one more possibility for your inspiration, motivated by my aversion to too long names:
`at_time_step(lag=i)` and acordingly for iteration, where `i=0` gives the current and `i=1` gives the previous one. This allows to unify with current time steps/iterates as well, e.g., `is_at_iteration(lag=i)` removes the need for splitting up the boolean methods (I could not find those in the code - maybe I misunderstood and these are not methods...)

Maybe there is a better word than `lag`. Yet, I wanted to consisely indicate the backward nature of what we use to call `index`, which is counterintuitive considering how one would write time discretizations and iterative methods on paper. In general, now having looked at the code, and with the above. Such a change would strongly suggest also renaming `time_step_index` to `time_step_lag`.

To make it even more complicated, one could provide two possibilities to access the operators and also `get_solution_values`, using not only indices of type `int`, but a `Literal["current", "previous"], e. g., `at_iterate("current")`, `at_iterate("previous")`, `get_solution_values(time_step="current"` etc. Of course a nightmare to maintain. But as your main motivation is expressivity of the code, my answer would look like that.

I anticipate that this is not of interest. Nevertheless, since I was blank last time I was asked and these thoughts now popped into my head, here it is: my brain dump :)

keileg commented:
@vlipovac @IvarStefansson, @jwboth: It seems to me that, while we agree the current `previous_timestep`, `previous_iterate` should be renamed, we are not ready to conclude on a naming scheme. Still, Veljko is right we should finalize this process so as not to block the merges for compositional flow. I therefore suggest we revert the naming to what it was before this PR and revisit at a later stage. From what I understand, such a resolution will allow us to merge this PR with little extra effort.

Does this make sense to you?

IvarStefansson commented:
> @vlipovac @IvarStefansson, @jwboth: It seems to me that, while we agree the current `previous_timestep`, `previous_iterate` should be renamed, we are not ready to conclude on a naming scheme. Still, Veljko is right we should finalize this process so as not to block the merges for compositional flow. I therefore suggest we revert the naming to what it was before this PR and revisit at a later stage. From what I understand, such a resolution will allow us to merge this PR with little extra effort.
> 
> Does this make sense to you?

Yes, I think this is a good solution!

vlipovac commented:
> > @vlipovac @IvarStefansson, @jwboth: It seems to me that, while we agree the current `previous_timestep`, `previous_iterate` should be renamed, we are not ready to conclude on a naming scheme. Still, Veljko is right we should finalize this process so as not to block the merges for compositional flow. I therefore suggest we revert the naming to what it was before this PR and revisit at a later stage. From what I understand, such a resolution will allow us to merge this PR with little extra effort.
> > Does this make sense to you?
> 
> Yes, I think this is a good solution!

It's a bit of a pity that we failed to include this trivial change here, are you sure about this?
To my understanding, the target revision to return to would be [6016bf2](https://github.com/pmgbergen/porepy/pull/1172/commits/6016bf2fc329fc3e111deac96fdd14beac7eee20), correct?

Let me just clarify on thing regarding the booleans:
They are decorated with `@property`, hence not callable. The value is computed based on the index value, and it's `True` for *all* instances at some index, other than the current one.
Implementing a callable check by passing an index, makes to me little sense, primarily because this is a duplicate functionality (the user can simply check the value of the index, which is a public property of these operators)

keileg commented:
> It's a bit of a pity that we failed to include this trivial change here, are you sure about this?

I agree it would be better to make a change we are all happy with now and be done with it. Problem is, we have not yet managed to decide precisely what we want to do, despite a week of discussions. My experience in such situations is, we can go on until we reach a point of exhaustion and make a decision that has a high risk of being suboptimal. The better option usually is to cut the losses for now, and revisit the topic at a later point, so that is what I suggest we do.

> To my understanding, the target revision to return to would be [6016bf2](https://github.com/pmgbergen/porepy/pull/1172/commits/6016bf2fc329fc3e111deac96fdd14beac7eee20), correct?

Yes.

vlipovac commented:
> I agree it would be better to make a change we are all happy with now and be done with it. Problem is, we have not yet managed to decide precisely what we want to do, despite a week of discussions. My experience in such situations is, we can go on until we reach a point of exhaustion and make a decision that has a high risk of being suboptimal. The better option usually is to cut the losses for now, and revisit the topic at a later point, so that is what I suggest we do.

Reverted the changes two commits back.

**For future reference**

Reverted commits with changes in naming:

[051e645](https://github.com/pmgbergen/porepy/pull/1172/commits/051e6452b9368061132f6bcefb6cb5cf3d8c0752)
[7cc6c22](https://github.com/pmgbergen/porepy/pull/1172/commits/7cc6c2267e4f52b78488a78c30ec87cb3d528931)
